src/sentry/monitors/logic/mark_ok.py

```diff
@@ -33,17 +21,40 @@ def mark_ok(checkin: MonitorCheckIn, ts: datetime):
         "next_checkin": next_checkin,
         "next_checkin_latest": next_checkin_latest,
     }
-    if checkin.status == CheckInStatus.OK:
-        if monitor_env.monitor.status != ObjectStatus.DISABLED:
-            params["status"] = MonitorStatus.OK
-        # in the future this will auto-resolve associated issues
-        if monitor_env.status != MonitorStatus.OK:
+
+    recovery_threshold = monitor_env.monitor.config.get("recovery_threshold", 0)
+
+    # The status will not be updated if they have an incident and the incident
+    # is not recovering
+    allow_status_update = True
+
+    # Resolve the incident if we have met the recovery recovery_threshold
+    if recovery_threshold:
+        previous_checkins = MonitorCheckIn.objects.filter(monitor_environment=monitor_env).order_by(
+            "-date_added"
+        )[:recovery_threshold]
+
+        # Incident recovers when ew have successive threshold check-ins
+        incident_recovering = all(
+            previous_checkin.status == CheckInStatus.OK for previous_checkin in previous_checkins
+        )
+
+        # Resolve the incident
+        if incident_recovering and monitor_env.status != MonitorStatus.OK:
```

Comment:
could we simplify the logic/execution overall here? Right now it seems like the order of conditions are:

1. Check if we have a recovery threshold (if not just update the status)
2. If we have, fetch N recent check-ins and check that they are all ok
3. If they are all ok check to see if our monitor is in an incident (!= MonitorStatus.OK)
4. If true, resolve the incident, and then allow a status update on the monitor

But would it be better to instead

1. Check `monitor_env.status != MonitorStatus.OK` first, if it IS `OK` then we can do nothing
2. If not `OK` then chcek for a recovery threshold, proceed with the previous steps 2 and then 4

Only mentioning this because hopefully the majority of the time, user's monitor statuses will be OK which means we shouldn't check the N most recent check-ins every time they send an OK check-in (assuming they have a recovery threshold)